rasa/graph_components/validators/config_validator.py

```diff
@@ -0,0 +1,532 @@
+from rasa.shared.importers.importer import TrainingDataImporter
+from rasa.shared.nlu.training_data.training_data import TrainingData
+import rasa.shared.utils.io
+
+
+# FIXME: check right paths, not just existence on any path (e.g. end2end featurization
+# might be separate from NLU featurization and only one of them might have a
+# tokenizer)
+
+TRAINABLE_EXTRACTORS = {
+    MitieEntityExtractorGraphComponent,
+    CRFEntityExtractorGraphComponent,
+    DIETClassifierGraphComponent,
+}
+
+# TODO: do we have a registries for these?
+FEATURIZER_CLASSES = {
+    LexicalSyntacticFeaturizerGraphComponent,
+    RegexFeaturizerGraphComponent,
+    # TODO: add rest / replace
+}
+POLICY_CLASSSES = {  # Note: this is used in tests only, should belong elsewhere anyway
+    TEDPolicyGraphComponent,
+    MemoizationPolicyGraphComponent,
+    RulePolicyGraphComponent,
+}
+
+
+def types_to_str(types: Iterable[Type]) -> Text:
+    """Returns a text containing the names of all given types.
+
+    Args:
+        types: some types
+    Returns:
+        text containing all type names
+    """
+    return ", ".join([type.__name__ for type in types])
+
+
+class ConfigValidator(GraphComponent):
+    """Validates the current graph schema against the training data and domain."""
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> ConfigValidator:
+        """Creates a new `ConfigValidator` (see parent class for full docstring)."""
+        return cls(execution_context.graph_schema)
+
+    def __init__(self, graph_schema: GraphSchema) -> None:
+        """Instantiates a new `ConfigValidator`.
+
+        Args:
+           graph_schema: a graph schema
+        """
+        self._graph_schema = graph_schema
+        self._component_types = set(node.uses for node in graph_schema.nodes.values())
+        self._policy_schema_nodes: List[SchemaNode] = [
+            node
+            for node in self._graph_schema.nodes.values()
+            if issubclass(node.uses, PolicyGraphComponent)
+        ]
+
+    def validate(self, importer: TrainingDataImporter) -> None:
+        """Validates the current graph schema against the training data and domain.
+
+        Args:
+            importer: the training data importer which can also load the domain
+        """
+        nlu_data = importer.get_nlu_data()
+        self._validate_nlu(nlu_data)
+
+        story_graph = importer.get_stories()
+        domain = importer.get_domain()
+        self._validate_core(story_graph, domain)
+
+    def _validate_nlu(self, training_data: TrainingData) -> None:
+        """Validates whether the configuration matches the training data.
+
+        Args:
+           training_data: training_data
+        """
+        # TODO: raise if training data is empty?
+        # TODO: where do we check whether NLU training data is consistent with
```

Comment:
I think nowhere as NLU historically didn't use a `Domain`. As Core/NLU more and more merge I think such as check would definitely make sense to implement at some point 👍🏻